### PR TITLE
FEAT-#4619: Integrate mypy static type checking

### LINKED ISSUE #4619: Integrate mypy static type checking
We want to follow up on #74 to add type hints to the Modin source code. #3899 discusses the idea of using pyright as our tool for static type checking, but I propose that we make use of mypy instead. The main reasons are that mypy has a larger development community, more widespread industry adoption, and better tools to scale with larger codebases. 

I propose that we integrate mypy over a series of steps:

1. Add a simple, configurable `mypy.ini` file that only checks a subset of files
2. Make according type checks for the subset of files
3. Add mypy checks to the precommit hooks
4. Add mypy checks to CI

From here, we can start to add more type checks to the rest of the code base (perhaps at a module by module basis).

### PR DESCRIPTION
**Pull request by pyrito**
## What do these changes do?
This PR addresses #4619 and implements the first cut of mypy integration. The goal is to start off with small modules and work our way up to the majority of our codebase. I decided to start off with the logging module since the amount of code is fairly small. Please note that this PR is still in the works as I need to add CI integration and make other changes. 

- [x] commit message follows format outlined [here](https://modin.readthedocs.io/en/latest/development/contributing.html#commit-message-formatting)
- [x] passes `flake8 modin/ asv_bench/benchmarks scripts/doc_checker.py`
- [x] passes `black --check modin/ asv_bench/benchmarks scripts/doc_checker.py`
- [x] signed commit with `git commit -s` 
- [x] Resolves #4619 
- [x] tests added and passing
- [x] module layout described at `docs/development/architecture.rst` is up-to-date 
- [x] added (Issue Number: PR title (PR Number)) and github username to release notes for next major release

### CODE REVIEW

File: modin/logging/config.py [hunk 1]
```diff
@@ -83,14 +86,17 @@ def bytes_int_to_str(num_bytes, suffix="B"):
         Human-readable string format.
     """
     factor = 1000
+    bytes: float = num_bytes
```
Comment (mvashishtha) [inline modin/logging/config.py:89]: `bytes` is a built-in keyword, so we shouldn't use it.

File: modin/logging/logger_decorator.py [hunk 2]
```diff
@@ -79,10 +79,10 @@ def enable_logging(
     log_level = log_level.lower()
     assert hasattr(Logger, log_level.lower()), f"Invalid log level: {log_level}"
 
-    def decorator(obj):
+    def decorator(obj: Any) -> Any:
         """Decorate function or class to add logs to Modin API function(s)."""
         if isinstance(obj, type):
-            seen = {}
+            seen: Dict[Any, Any] = {}
```
Comment (mvashishtha) [inline modin/logging/logger_decorator.py:85]:
```suggestion
            seen: Dict[str, Any] = {}
```

Is this correct?

File: modin/logging/logger_decorator.py [hunk 3]
```diff
@@ -103,11 +103,13 @@ def decorator(obj):
         elif isinstance(obj, staticmethod):
             return staticmethod(decorator(obj.__func__))
 
+        assert isinstance(modin_layer, str)
```
Comment (mvashishtha) [inline modin/logging/logger_decorator.py:106]: If we assert that `modin_layer` is a string, why does the typing allow it not to be a string?

File: mypy.ini [hunk 4]
```diff
@@ -0,0 +1,23 @@
+[mypy]
+python_version = 3.9
```
Comment (mvashishtha) [inline mypy.ini:2]: Why 3.9? I know python 3.10 currently doesn't work with Modin. Will we have to update this?

File: mypy.ini [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+follow_imports = silent
+
+# be strict
+disallow_untyped_calls=True
+disallow_untyped_defs=True
+strict_optional=True
+warn_no_return=True
+warn_redundant_casts=True
+warn_unused_ignores=True
+# Should we disallow generics?
```
Comment (mvashishtha) [inline mypy.ini:17]: I think not.

File: contributing/pre-commit [hunk 6]
```diff
@@ -11,4 +11,7 @@ printf "running flake8. This script will preempt the commit if flake8 fails.\n"
 flake8 modin/ asv_bench/benchmarks scripts/doc_checker.py
 printf "flake8 passed!\n"
 
+printf "running mypy. This script will preempt the commit if mypy fails.\n"
```
Comment (mvashishtha) [inline contributing/pre-commit:14]: I think I'm the only person who uses this hook :) We should add the MyPy check to [CI](https://github.com/modin-project/modin/blob/3982306a540bd09b43a21dcdca6b16b4741ed151/.github/workflows/ci.yml). Is there a way to have it check all changed lines in other files, in addition to all of `**/modin/logging/*.py`?

File: modin/logging/config.py (hunk 1 above)
Comment (pyrito) [reply modin/logging/config.py:89]: This is a good point, I'll make the change

File: modin/logging/logger_decorator.py (hunk 3 above)
Comment (pyrito) [reply modin/logging/logger_decorator.py:106]: There's a bit of complexity here. Basically `enable_logging` returns a function since it's a decorator. If you see in the top of the function, we check to see whether `modin_layer` is a string or not. If it's not, then we call `enable_logging()(modin_layer)` which is basically doing `decorator(modin_layer)`. We can see in the implementation of `decorator` that we check to see what the type of `obj` is (and we claim earlier that it can be a string, type, or callable function). We check for these cases but I think `mypy` isn't convinced that we can get away with calling string methods on the union, so I added an assert for sanity. Does that make sense?

Comment (pyrito) [reply modin/logging/logger_decorator.py:106]: mypy threw this error specifically:
```
modin/logging/logger_decorator.py:106:32: error: Item "function" of "Union[str, Callable[..., Any], Type[Any]]" has no attribute "upper"  [union-attr]
modin/logging/logger_decorator.py:107:30: error: Item "function" of "Union[str, Callable[..., Any], Type[Any]]" has no attribute "upper"  [union-attr]
```

File: contributing/pre-commit (hunk 6 above)
Comment (mvashishtha) [reply contributing/pre-commit:14]: But we should also keep this change! I know I will want to add this to my pre-commit hook.

File: mypy.ini (hunk 4 above)
Comment (pyrito) [reply mypy.ini:2]: Reading the docs online, it looks like mypy will default to using whatever python version is used to run `mypy` itself. I set 3.9 because it seems it doesn't accept minor versions (e.g. 3.9.12) and I wasn't sure about 3.10. I could just get rid of this, but I figured that pinning a version is always better than not doing so.

File: contributing/pre-commit (hunk 6 above)
Comment (jeffreykennethli) [reply contributing/pre-commit:14]: I use this hook too!

Comment (pyrito) [reply contributing/pre-commit:14]: Haha, I actually use this pre-commit hook. It's super useful! And yes, I intend on adding the MyPy check to CI. I'm not sure if we can have it check all changed lines in other files. I think it would be hard for mypy to do it only on recent diffs. I think we're going to have to slowly make type changes over time, but I'll try to see if I can find anything online about this.

File: modin/logging/logger_decorator.py (hunk 2 above)
Comment (pyrito) [reply modin/logging/logger_decorator.py:85]: I don't think so. If you see the usage of `seen` it seems to be looking things up with `attr_value` as the key, and the type for that seems to be `Any`.

File: modin/logging/config.py [hunk 7]
```diff
@@ -37,7 +38,9 @@
 class ModinFormatter(logging.Formatter):  # noqa: PR01
     """Implement custom formatter to log at microsecond granularity."""
 
-    def formatTime(self, record, datefmt=None):
+    def formatTime(
+        self, record: logging.LogRecord, datefmt: Optional[str] = None
+    ) -> str:
```
[the remark this answers is not in the document]
Comment (pyrito) [reply modin/logging/config.py:43]: Yes, I was going to get to this. I was thinking of making a separate PR or I could just roll it into this one?

File: modin/logging/config.py [hunk 8]
```diff
@@ -83,14 +86,17 @@ def bytes_int_to_str(num_bytes, suffix="B"):
...
-    return f"{num_bytes * 1000:.2f}P{suffix}"
+        if bytes < factor:
+            return f"{bytes:.2f}{unit}{suffix}"
+        bytes /= factor
+    return f"{bytes * 1000:.2f}P{suffix}"
 
 
-def _create_logger(namespace, job_id, log_name, log_level):
+def _create_logger(
+    namespace: str, job_id: str, log_name: str, log_level: int
```
[the remark this answers is not in the document]
Comment (pyrito) [reply modin/logging/config.py:98]: We would need to change some of the code to go from using the enum to a string. Is that a change we want to make?

File: modin/logging/logger_decorator.py (hunk 3 above)
Comment (pyrito) [reply modin/logging/logger_decorator.py:106]: @mvashishtha would it make sense to add a string to give more context on a potential failure on this assert here?

Comment (mvashishtha) [reply modin/logging/logger_decorator.py:106]: Yeah, I think a short comment would be good.

File: modin/logging/logger_decorator.py (hunk 2 above)
Comment (vnlitvinov) [reply modin/logging/logger_decorator.py:85]: It should really be `Dict[Callable, Callable]`, as it's mapping unwrapped function to wrapped function, to handle cases like
```python
@enable_logging
class Foo:
    def bar(self):
        pass
    qux = bar

assert Foo.bar == Foo.qux
```

File: modin/logging/logger_decorator.py (hunk 3 above)
Comment (vnlitvinov) [reply modin/logging/logger_decorator.py:106]: I would advise against using bare `assert`-s, I think it's better to accompany each such statement with a proper assertion message which would double-serve as a comment

File: mypy.ini (hunk 4 above)
Comment (vnlitvinov) [reply mypy.ini:2]: I would rather we don't pin the version, as pinned version prohibits from running the checks with whatever version it is installed.

As for 3.10 - I think Modin should support that (we should be pretty agnostic on upper bound, requiring >=3.8), it's Ray which doesn't support 3.10 now.

File: requirements-dev.txt [hunk 9]
```diff
@@ -46,3 +46,6 @@ flake8
 numpydoc==1.1.0
 # experimental version of fuzzydata requires at least 0.0.6 to successfully resolve all dependencies
 fuzzydata>=0.0.6
+# Pin versions of mypy and pandas stubs for stability
+mypy==0.961
+pandas-stubs==1.4.2.220626
```
Comment (vnlitvinov) [inline requirements-dev.txt:51]: we should now have a doc saying "how to update pandas version", as we have a bunch of different places where it is mentioned

**Comment by vnlitvinov** [comment]
Unless `mypy` is run by CI, it won't protect anything really. Please add a CI check for that!

File: modin/logging/config.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (pyrito) [reply modin/logging/config.py:98]: I think we would also have to reflect the changes on all the callers of the function. Is that something. we want to do as well?

File: modin/logging/logger_decorator.py (hunk 2 above)
Comment (pyrito) [reply modin/logging/logger_decorator.py:85]: I had tried that before but mypy complained, so I kind of copped out to use the `Any`:
```
modin/logging/logger_decorator.py:91:40: error: Invalid index type "Union[FunctionType, MethodType, classmethod[Any], staticmethod[Any]]" for "Dict[Callable[..., Any], Callable[..., Any]]"; expected type "Callable[..., Any]"  [index]
modin/logging/logger_decorator.py:93:40: error: Invalid index type "Union[FunctionType, MethodType, classmethod[Any], staticmethod[Any]]" for "Dict[Callable[..., Any], Callable[..., Any]]"; expected type "Callable[..., Any]"  [index]
```

File: .github/workflows/ci.yml [hunk 10]
```diff
@@ -27,6 +27,20 @@ jobs:
...
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v2
+        with:
+          fetch-depth: 1
+      - uses: actions/setup-python@v2
+        with:
+          python-version: "3.8.x"
+          architecture: "x64"
+      - run: pip install -r requirements-dev.txt
```
Comment (mvashishtha) [inline .github/workflows/ci.yml:41]: Can we just `pip install mypy` as we do with `black` above?

File: environment-dev.yml [hunk 11]
```diff
@@ -36,6 +36,8 @@ dependencies:
   - scikit-learn
   - pymssql
   - psycopg2
+  - mypy==0.961
```
Comment (mvashishtha) [inline environment-dev.yml:39]: We usually try not to pin. Is there a reason to pin this version? I see we don't pin `black` or `flake8`.

File: modin/logging/config.py [hunk 12]
```diff
@@ -83,14 +86,17 @@ def bytes_int_to_str(num_bytes, suffix="B"):
...
+    n_bytes: float = num_bytes
     for unit in ["", "K", "M", "G", "T", "P"]:
-        if num_bytes < factor:
-            return f"{num_bytes:.2f}{unit}{suffix}"
-        num_bytes /= factor
-    return f"{num_bytes * 1000:.2f}P{suffix}"
+        if n_bytes < factor:
+            return f"{n_bytes:.2f}{unit}{suffix}"
+        n_bytes /= factor
+    return f"{n_bytes * 1000:.2f}P{suffix}"
```
Comment (naren-ponder) [inline modin/logging/config.py:94]: Why is this change necessary? Can we instead make `num_bytes` a float at the function signature and change the docstring? I'm just not a fan of `n_bytes: float = num_bytes`.

File: .github/workflows/ci.yml (hunk 10 above)
Comment (pyrito) [reply .github/workflows/ci.yml:41]: I decided to install all of it because I think `mypy` might require all packages to be installed so that it can do the type checking.

File: modin/logging/config.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (jeffreykennethli) [reply modin/logging/config.py:98]: Yes, but no need to do any of this in this PR, this was more of a general comment. If you think it's worth changing, feel free to open an issue for it.

File: environment-dev.yml (hunk 11 above)
Comment (pyrito) [reply environment-dev.yml:39]: I pinned the version to avoid any kind of surprises but I guess there's no reason to do so. I can get rid of the version pinning.

**Review by jeffreykennethli — APPROVED** [review]

File: modin/logging/config.py (hunk 12 above)
Comment (pyrito) [reply modin/logging/config.py:94]: If I were to leave it as is, I'd get an error like this:
```
modin/logging/config.py:93:9: error: Incompatible types in assignment (expression has type "float", variable has type "int")  [assignment]
```

One alternative solution might be to do something like: 
```python
n_bytes = num_bytes # type: float
```
This is a little janky workaround and seems like a semantic difference. Maybe if we wanted to be more explicit we could do something like:
```python
n_bytes = float(num_bytes)
```

Does anyone else have any thoughts about this? @modin-project/modin-contributors

Comment (naren-ponder) [reply modin/logging/config.py:94]: Agreed with @mvashishtha.

File: modin/logging/config.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (pyrito) [reply modin/logging/config.py:99]: Yeah looking at it now, I think that we should probably unify the log_levels to either be `str` or `int` or maybe even both. I'll create an issue for this. It might be a good one for new folks to try working on.
